### DISCUSSION TITLE
Apply dithering to dimmed and blurred backgrounds

### DISCUSSION THREAD

**Discussion by AuracleTech** [ppy/osu#17668] [Ideas]
When using a high dim level and blur the colors are banding ( as they should )
I propose a new option that would make the blur frosted, removing the background color banding!
![image](https://user-images.githubusercontent.com/40531871/161826348-0442a7cb-09a3-4e97-9da7-76bd5d492885.png)

Added brightness for those who can't see the difference on their specific monitor.
![image](https://user-images.githubusercontent.com/40531871/161991275-980774fb-04fc-4b5b-ba19-325a10bed669.png)

**Comment by thacer13** [comment]
Yep, banding can look really bad depending on how your monitor is calibrated.

**Reply by AuracleTech** [reply]
That's not how banding works.
Of course banding is worst on bad monitors, but banding is an inevitable issue.
But there's small hacks like frosting that will improve it!
Banding is caused by the lack of color range, so if you lower an image brightness and blur it, the color space will be incredibly narrow, so dispersing the pixels can be a workaround and make it appear with less bands.

**Reply by thacer13** [reply]
Not saying it's a 'fixable issue', I just noticed that it's more apparent in some monitors/configs than in others, and the method you proposed looks nice.

**Reply by AuracleTech** [reply]
Indeed!
On my cheap 4K tv the banding is incredibly apparent on my main monitor its almost not visible, but still very visible at high dimming.

**Comment by peppy** [comment]
The "frosted" version in your proposal has dithering applied. I'm not sure what you mean by frosted, but dithering may be hard to make work, while not impossible.

**Reply by AuracleTech** [reply]
Exactly.
I have no idea on how hard it would be to implement it / how performance heavy it would be.
But TaskbarX is using it so I figured maybe its doable, I honestly have no idea, I just thought it looked nicer.

https://user-images.githubusercontent.com/40531871/161882472-b57bee40-4832-408b-89d2-b4b5c9141a46.mp4

**Reply by AuracleTech** [reply]
I made this a while ago, it just uses a simple noise texture, this could done rather easily in game
[Dithering using texture](https://codepen.io/Auracle/pen/eYzpjro)

**Reply by smoogipoo** [reply]
Very simple implementation based on the above would look something like this:
```diff
diff --git a/osu.Game/Graphics/Backgrounds/BeatmapBackground.cs b/osu.Game/Graphics/Backgrounds/BeatmapBackground.cs
index e0c15dd52a..25fe6f9b0a 100644
--- a/osu.Game/Graphics/Backgrounds/BeatmapBackground.cs
+++ b/osu.Game/Graphics/Backgrounds/BeatmapBackground.cs
@@ -2,6 +2,10 @@
 // See the LICENCE file in the repository root for full licence text.

 using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.OpenGL.Textures;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
 using osu.Framework.Graphics.Textures;
 using osu.Game.Beatmaps;

@@ -23,6 +27,21 @@ public BeatmapBackground(WorkingBeatmap beatmap, string fallbackTextureName = @"
         private void load(LargeTextureStore textures)
         {
             Sprite.Texture = Beatmap?.Background ?? textures.Get(fallbackTextureName);
+
+            var noiseTexture = textures.Get("https://i.ibb.co/mFpMBnQ/perfnoise.png", WrapMode.Repeat, WrapMode.Repeat);
+
+            AddInternal(new Sprite
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+                RelativeSizeAxes = Axes.Both,
+                FillMode = FillMode.Fill,
+                Texture = noiseTexture,
+                TextureRectangle = new RectangleF(0, 0, noiseTexture.Width, noiseTexture.Height),
+                TextureRelativeSizeAxes = Axes.None,
+                Alpha = 0.28f,
+                Depth = -1
+            });
         }

         public override bool Equals(Background other)
```
Wouldn't be too hard to do shader-level dithering either.

**Reply by peppy** [reply]
would definitely need to be shader level. we shouldn't be doubling fill rate to achieve this. also a texture is kinda faking it in a way that could look bad in some cases.

**Reply by AuracleTech** [reply]
I made this for my game engine to fix banding and I remembered this thread
```glsl
float dither(vec2 uv)
{
    return fract(sin(dot(uv.xy, vec2(12.9898, 78.233))) * 43758.5453) / 256.0 - 0.001953125;
}
```
`result += vec4(vec3(dither(tex_coord)), 1.0);`
[image requires full resolution for render-like result](https://user-images.githubusercontent.com/40531871/228197729-880ba233-c495-41bb-94e8-a73574de222c.png)
